# Where to eat in Chigago : an analysis of the inspections from the Chicago department of Public Health's Food Protection Programm

## 1. Introduction

The Chicago department of Public Health’s Food Protection Program provides a database which contains the information from inspection reports of restaurants and other food establishments in Chicago from 2010 to the present. It contains many informations about the establishments, like their type of facility (groceries’ stores, restaurants, coffee shop, …) and their locations. Many informations about the violations listed are also provided in the database, like the description of the findings that caused them and the reason that invoked the staff from the program to led an inspection.

In our project we endeavor to visualize the healthiness of public food establishments according to their type of facility, their ward and the date of the inspection. An analysis of the violation’s types according to these three parameters will also be conducted. 

The principal questions we'll answer are : 
    - Which ward of Chicago are the most healthy and unhealthy ? 
    - Which type of facility tend to be less healthy ? 
    - Did the healthiness of the food in Chicago increase or decrease from 2010 until now ?

New problematics could be asked during the analysis and would be added to these.

The purpose of the project is to help the consumer to easily choose where to eat in Chicago and to provide them an interactive and intuitive way to browse the different places offered to them. Also, it could help the Chicago department of Public Health’s Food Protection Program to adapt their methods relying on the situation described by the findings of the analysis (for example, if a prevention program should be proposed for a specific area or type of facility).

In [1]:
%matplotlib inline
import pandas as pd
import numpy as np
import re
from matplotlib.ticker import MaxNLocator
import matplotlib.pyplot as plt
import requests as req
from bs4 import BeautifulSoup
import seaborn as sns

## 2. Preprocessing

### 2.1 Facilities of interest Selection

First a quick look at how is organized the dataset. 

In [2]:
df = pd.read_csv('food-inspections.csv',sep=',') #creation of the dataframe
df.head(3)

Inspection ID                          DBA Name  \
0        2320971  JUMPSTART EARLY LEARNING ACADEMY   
1        2320918                         BEEFSTEAK   
2        2320986                     BABA'S COFFEE   

                           AKA Name  License #                 Facility Type  \
0  JUMPSTART EARLY LEARNING ACADEMY  2589822.0  Children's Services Facility   
1                         BEEFSTEAK  2698445.0                    Restaurant   
2                     BABA'S COFFEE  2423353.0                    Restaurant   

            Risk                  Address     City State      Zip  ...  \
0  Risk 1 (High)       7559 W ADDISON ST   CHICAGO    IL  60634.0  ...   
1  Risk 1 (High)       303 E SUPERIOR ST   CHICAGO    IL  60611.0  ...   
2  Risk 1 (High)  5544-5546 N KEDZIE AVE   CHICAGO    IL  60625.0  ...   

              Results                                         Violations  \
0  Pass w/ Conditions  5. PROCEDURES FOR RESPONDING TO VOMITING AND D...   
1                Pass  39. CONTAMINATION PREVENTED DURING FOOD PREPAR...   
2            No Entry                                                NaN   

    Latitude  Longitude                                           Location  \
0  41.945065 -87.816734  {'longitude': '41.945064857019986', 'latitude'...   
1  41.895692 -87.620143  {'longitude': '41.895692401410514', 'latitude'...   
2  41.982582 -87.708996  {'longitude': '41.98258181784537', 'latitude':...   

   Historical Wards 2003-2015 Zip Codes  Community Areas  Census Tracts  Wards  
0                         NaN       NaN              NaN            NaN    NaN  
1                         NaN       NaN              NaN            NaN    NaN  
2                         NaN       NaN              NaN            NaN    NaN  

[3 rows x 22 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195312 entries, 0 to 195311
Data columns (total 22 columns):
Inspection ID                 195312 non-null int64
DBA Name                      195312 non-null object
AKA Name                      192862 non-null object
License #                     195295 non-null float64
Facility Type                 190535 non-null object
Risk                          195239 non-null object
Address                       195312 non-null object
City                          195173 non-null object
State                         195270 non-null object
Zip                           195261 non-null float64
Inspection Date               195312 non-null object
Inspection Type               195311 non-null object
Results                       195312 non-null object
Violations                    143530 non-null object
Latitude                      194627 non-null float64
Longitude                     194627 non-null float64
Location                      194627 n

It is a dataset of 195'312 entries with 22 columns listed above.
First thing first, we want to put the different facility types in categories that make sense for our project.

In [4]:
#df['Facility Type'].unique()

This command returns a list which contains the different facility types found in the `Facility Type column`. A lot of different types of facility are found in the data.

First, we thought about only select the "private" establishments, where it is possible to eat a main course (for example, the places where you can only eat an ice cream are deleted of our list). They all are categorized in order to be compared with each other.


In [5]:
public_dic = {'restaurant' : ['Restaurant', 'DINING HALL', 'TENT RSTAURANT'], \
              'grocery_restaurant' : ['Grocery & Restaurant', 'GROCERY& RESTAURANT', 'GROCERY/RESTAURANT',\
                                    'GROCERY/ RESTAURANT', 'GROCERY STORE/ RESTAURANT', 'GROCERY & RESTAURANT',\
                                    'RESTAURANT/GROCERY', 'grocery & restaurant', 'RESTAURANT/GROCERY STORE',\
                                    'GROCERY/TAQUERIA', 'GAS STATION/RESTAURANT'],\
              'banquet' : ['LOUNGE/BANQUET HALL', 'BANQUET', 'Banquet Hall', 'BANQUET FACILITY', 'banquet hall',\
                         'banquets', 'Banquet Dining',  'Banquet/kitchen','RESTAURANT.BANQUET HALLS',\
                         'BANQUET HALL', 'Banquet', 'BOWLING LANES/BANQUETS'], \
              'rooftop_restaurant' : ['Wrigley Roof Top', 'REST/ROOFTOP'],\
              'bar_restaurant' : ['RESTAURANT/BAR', 'RESTUARANT AND BAR', 'BAR/GRILL', 'RESTAURANT/BAR/THEATER',\
                                'JUICE AND SALAD BAR', 'SUSHI COUNTER', 'TAVERN/RESTAURANT', 'tavern/restaurant',\
                                'TAVERN GRILL'], \
              'bakery_restaurant' : ['BAKERY/ RESTAURANT', 'bakery/restaurant', 'RESTAURANT/BAKERY'], \
              'liquor_restaurant' : ['RESTAURANT AND LIQUOR', 'RESTAURANT/LIQUOR'], \
              'catering' : ['CATERING/CAFE', 'Catering'], \
              'golden_diner' : ['Golden Diner']}

In [6]:
facilitytype = 'BANQUET'
len(df[df['Facility Type'] == facilitytype])

64

This command returns the number of occurencs of the `Facility Type` inputed. 

With trying different types previously categorized and listed in the `public_dic` dictionary we have noted that the results were too distant to conduct a meaningful analysis. That's why we then decided to also select "public" establishments like school cafeterias and hospitals. It could be interesting to compare private and public inspection results.


In [7]:
private_dic = {'daycare' : ['Daycare Above and Under 2 Years', 'Daycare (2 - 6 Years)', 'Daycare Combo 1586',\
                          'Daycare (Under 2 Years)', 'DAYCARE 2 YRS TO 12 YRS', 'Daycare Night', 'DAY CARE 2-14',\
                          'Daycare (2 Years)', 'DAYCARE', 'ADULT DAYCARE', '15 monts to 5 years old', 'youth housing',\
                          'DAYCARE 1586', 'DAYCARE COMBO', '1584-DAY CARE ABOVE 2 YEARS', 'CHURCH/DAY CARE', 'DAY CARE',\
                          'DAYCARE 6 WKS-5YRS', 'DAY CARE 1023', 'DAYCARE 2-6, UNDER 6', 'Day Care Combo (1586)'], \
               'school' : ['SCHOOL', 'School', 'PRIVATE SCHOOL', 'AFTER SCHOOL PROGRAM', 'COLLEGE',\
                         'BEFORE AND AFTER SCHOOL PROGRAM', 'Private School', 'TEACHING SCHOOL',\
                         'PUBLIC SHCOOL', 'CHARTER SCHOOL CAFETERIA', 'CAFETERIA', 'Cafeteria', 'cafeteria',\
                         'UNIVERSITY CAFETERIA', 'PREP INSIDE SCHOOL', 'CHARTER SCHOOL', 'school cafeteria',\
                         'CHARTER SCHOOL/CAFETERIA', 'School Cafeteria', 'ALTERNATIVE SCHOOL', 'CITY OF CHICAGO COLLEGE',\
                         'after school program', 'CHURCH/AFTER SCHOOL PROGRAM', 'AFTER SCHOOL CARE'], \
               'childrens_services' : ["Children's Services Facility", 'CHILDRENS SERVICES FACILITY', \
                                     "CHILDERN'S SERVICE FACILITY", "1023 CHILDREN'S SERVICES FACILITY", \
                                     "1023 CHILDERN'S SERVICES FACILITY", "1023-CHILDREN'S SERVICES FACILITY", \
                                     "1023 CHILDERN'S SERVICE FACILITY", "1023 CHILDERN'S SERVICE S FACILITY", \
                                     'CHILDERN ACTIVITY FACILITY', "CHILDERN'S SERVICES  FACILITY", '1023'], \
               'adultcare' : ['Long Term Care', 'REHAB CENTER', 'Hospital', 'ASSISTED LIVING', 'SENIOR DAY CARE',\
                            'Assisted Living', 'NURSING HOME', 'ASSISTED LIVING FACILITY', 'SUPPORTIVE LIVING FACILITY',\
                            'Assisted Living Senior Care', 'Adult Family Care Center', '1005 NURSING HOME', \
                            'Long-Term Care Facility', 'LONG TERM CARE FACILITY', 'ASSISSTED LIVING',\
                            'Long-Term Care','Long Term Care Facility', 'VFW HALL']}

In [8]:
total_dic = {**public_dic , **private_dic}

In [9]:
def newcolfromdict(dataf, dic, inputcolumn, outputcolumn) :
    new_list = []
    for content in dataf[inputcolumn] :
        new_content = 'Not listed'

        for k, v in dic.items() :
            if type(v) == list :
                for element in v :
                    if content == element :
                        new_content = k
            else :
                if content == v :
                    new_content = k

        new_list.append(new_content)
    dataf[outputcolumn] = new_list
    return dataf

The **newcolfromdict** function constructs a new column `outputcolum` from the inputed column `inputcolum` comparing its values to the `dic` dictionary and adds it to the `dataf` dataframe.

### 2.2 Columns Cleaning

### A) City Selection

To construct the new dataframe, the `Facility Type` column is dropped because it has been replaced by the column `Facility group` and the *Not Listed* establishments are not selected.

The duplicates are dropped.

In [10]:
eat_seat = newcolfromdict(df, total_dic, 'Facility Type', 'Facility Group')
eat_seat = df.loc[df['Facility Group'] != 'Not Listed']
eat_seat = eat_seat.drop(columns = ['Facility Type'])

eat_seat = eat_seat.drop_duplicates()

eat_seat.head(2)

Inspection ID                          DBA Name  \
0        2320971  JUMPSTART EARLY LEARNING ACADEMY   
1        2320918                         BEEFSTEAK   

                           AKA Name  License #           Risk  \
0  JUMPSTART EARLY LEARNING ACADEMY  2589822.0  Risk 1 (High)   
1                         BEEFSTEAK  2698445.0  Risk 1 (High)   

              Address     City State      Zip          Inspection Date  ...  \
0  7559 W ADDISON ST   CHICAGO    IL  60634.0  2019-11-01T00:00:00.000  ...   
1  303 E SUPERIOR ST   CHICAGO    IL  60611.0  2019-11-01T00:00:00.000  ...   

                                          Violations   Latitude  Longitude  \
0  5. PROCEDURES FOR RESPONDING TO VOMITING AND D...  41.945065 -87.816734   
1  39. CONTAMINATION PREVENTED DURING FOOD PREPAR...  41.895692 -87.620143   

                                            Location  \
0  {'longitude': '41.945064857019986', 'latitude'...   
1  {'longitude': '41.895692401410514', 'latitude'...   

   Historical Wards 2003-2015 Zip Codes  Community Areas  Census Tracts  \
0                         NaN       NaN              NaN            NaN   
1                         NaN       NaN              NaN            NaN   

   Wards      Facility Group  
0    NaN  childrens_services  
1    NaN          restaurant  

[2 rows x 22 columns]

Since we only care about establishments in Chicago Illinois, we will only keep the data for this city and drop the `City` and `State` columns.

We first check the different City's names to avoid deleting rows due to missprints.

In [11]:
df['City'].unique()

array(['CHICAGO', nan, 'chicago', 'Chicago', 'GRIFFITH', 'NEW YORK',
       'SCHAUMBURG', 'ELMHURST', 'ALGONQUIN', 'NEW HOLSTEIN', 'CCHICAGO',
       'NILES NILES', 'EVANSTON', 'CHICAGO.', 'CHESTNUT STREET',
       'LANSING', 'CHICAGOCHICAGO', 'WADSWORTH', 'WILMETTE', 'WHEATON',
       'CHICAGOHICAGO', 'ROSEMONT', 'CHicago', 'CALUMET CITY',
       'PLAINFIELD', 'HIGHLAND PARK', 'PALOS PARK', 'ELK GROVE VILLAGE',
       'CICERO', 'BRIDGEVIEW', 'OAK PARK', 'MAYWOOD', 'LAKE BLUFF',
       '312CHICAGO', 'SCHILLER PARK', 'SKOKIE', 'BEDFORD PARK',
       'BANNOCKBURNDEERFIELD', 'CHCICAGO', 'BLOOMINGDALE', 'Norridge',
       'CHARLES A HAYES', 'CHCHICAGO', 'CHICAGOI', 'SUMMIT',
       'OOLYMPIA FIELDS', 'WESTMONT', 'CHICAGO HEIGHTS', 'JUSTICE',
       'TINLEY PARK', 'LOMBARD', 'EAST HAZEL CREST', 'COUNTRY CLUB HILLS',
       'STREAMWOOD', 'BOLINGBROOK', 'INACTIVE', 'BERWYN', 'BURNHAM',
       'DES PLAINES', 'LAKE ZURICH', 'OLYMPIA FIELDS', 'alsip',
       'OAK LAWN', 'BLUE ISLAND', 'GLENCOE',

In [12]:
cities = ['CHICAGO','chicago','Chicago','CCHICAGO','CHICAGO.','CHICAGOCHICAGO','CHICAGOHICAGO',\
          'CHicago','312CHICAGO','CHCICAGO','CHCHICAGO','CHICAGOI','CHICAGO HEIGHTS']

eat_seat = eat_seat.loc[eat_seat['City'].isin(cities)]

eat_seat = eat_seat.drop(columns = ['City','State'])

# à la limite pour être stylés on pourrait tenter une expression régulière pour filtrer les différentes orthographes 
#de chigago au lieu de sélectionner à la main

Then we want to check the missing values.

In [13]:
eat_seat.isnull().sum()

Inspection ID                      0
DBA Name                           0
AKA Name                        2416
License #                         17
Risk                              70
Address                            0
Zip                                3
Inspection Date                    0
Inspection Type                    1
Results                            0
Violations                     51550
Latitude                         515
Longitude                        515
Location                         515
Historical Wards 2003-2015    194729
Zip Codes                     194729
Community Areas               194729
Census Tracts                 194729
Wards                         194729
Facility Group                     0
dtype: int64

In [14]:
print(len(eat_seat.index))    ##returns the number of rows of the df

194729


As we can see, the columns `Historical Wards 2003-2015`, `Zip Codes`, `Community Areas`, `Census Tracts` and `Wards` are empty and will be dropped.

We will only be using the `DBA Name` (the name under which the establishment is doing business ; DBA = doing business as), so we drop the `AKA Name` column too.

In [15]:
eat_seat = eat_seat.drop(columns = ['AKA Name','Historical Wards 2003-2015', 'Zip Codes', 'Community Areas',\
                                    'Census Tracts', 'Wards'])
eat_seat.head(2)

Inspection ID                          DBA Name  License #           Risk  \
0        2320971  JUMPSTART EARLY LEARNING ACADEMY  2589822.0  Risk 1 (High)   
1        2320918                         BEEFSTEAK  2698445.0  Risk 1 (High)   

              Address      Zip          Inspection Date Inspection Type  \
0  7559 W ADDISON ST   60634.0  2019-11-01T00:00:00.000         Canvass   
1  303 E SUPERIOR ST   60611.0  2019-11-01T00:00:00.000         License   

              Results                                         Violations  \
0  Pass w/ Conditions  5. PROCEDURES FOR RESPONDING TO VOMITING AND D...   
1                Pass  39. CONTAMINATION PREVENTED DURING FOOD PREPAR...   

    Latitude  Longitude                                           Location  \
0  41.945065 -87.816734  {'longitude': '41.945064857019986', 'latitude'...   
1  41.895692 -87.620143  {'longitude': '41.895692401410514', 'latitude'...   

       Facility Group  
0  childrens_services  
1          restaurant

### B) Inspection date and Risk

There is also adjustments to make in some columns, because the formats can be optimized :

* In `Inspection Date`, only the day will be kept, not the time of day that is actually not given
* In `Risk`, only the number will remain

For the column `Risk`, we first want to check what types of risk are listed.

In [16]:
eat_seat['Inspection Date'] = eat_seat['Inspection Date'].apply(lambda x:x.split('T')[0])

In [17]:
eat_seat.Risk.unique()
#print all types of entries in the column Risk

array(['Risk 1 (High)', 'Risk 2 (Medium)', 'Risk 3 (Low)', nan, 'All'],
      dtype=object)

We will replace **All** and **High Risk** by *3*, **Medium Risk** by *2* and **Low Risk** by *1*.


***--> pourquoi remplacer All by High Risk ?***

***A vérifier si c'était vraiment pensé comme ça quand la bdd a été conçue***

In [18]:
eat_seat['Risk'] = eat_seat['Risk'].replace({'All':1, 'Risk 1 (High)':3, 'Risk 2 (Medium)':2, 'Risk 3 (Low)':1})
eat_seat.head(2)

Inspection ID                          DBA Name  License #  Risk  \
0        2320971  JUMPSTART EARLY LEARNING ACADEMY  2589822.0   3.0   
1        2320918                         BEEFSTEAK  2698445.0   3.0   

              Address      Zip Inspection Date Inspection Type  \
0  7559 W ADDISON ST   60634.0      2019-11-01         Canvass   
1  303 E SUPERIOR ST   60611.0      2019-11-01         License   

              Results                                         Violations  \
0  Pass w/ Conditions  5. PROCEDURES FOR RESPONDING TO VOMITING AND D...   
1                Pass  39. CONTAMINATION PREVENTED DURING FOOD PREPAR...   

    Latitude  Longitude                                           Location  \
0  41.945065 -87.816734  {'longitude': '41.945064857019986', 'latitude'...   
1  41.895692 -87.620143  {'longitude': '41.895692401410514', 'latitude'...   

       Facility Group  
0  childrens_services  
1          restaurant

In [19]:
eat_seat = eat_seat.rename(columns={"License #": "License"}) #rename the column 'License #' into 'License'

In [20]:
len(eat_seat.License.unique())

37145

In [21]:
eat_seat.head()

Inspection ID                          DBA Name    License  Risk  \
0        2320971  JUMPSTART EARLY LEARNING ACADEMY  2589822.0   3.0   
1        2320918                         BEEFSTEAK  2698445.0   3.0   
2        2320986                     BABA'S COFFEE  2423353.0   3.0   
3        2320910           J.T.'S GENUINE SANDWICH  2689893.0   3.0   
4        2320904        KID'Z COLONY DAYCARE, INC.  2215609.0   3.0   

                   Address      Zip Inspection Date Inspection Type  \
0       7559 W ADDISON ST   60634.0      2019-11-01         Canvass   
1       303 E SUPERIOR ST   60611.0      2019-11-01         License   
2  5544-5546 N KEDZIE AVE   60625.0      2019-11-01         Canvass   
3       3970 N ELSTON AVE   60618.0      2019-11-01         License   
4       6287 S ARCHER AVE   60638.0      2019-11-01         Canvass   

              Results                                         Violations  \
0  Pass w/ Conditions  5. PROCEDURES FOR RESPONDING TO VOMITING AND D...   
1                Pass  39. CONTAMINATION PREVENTED DURING FOOD PREPAR...   
2            No Entry                                                NaN   
3                Pass  51. PLUMBING INSTALLED; PROPER BACKFLOW DEVICE...   
4                Fail  16. FOOD-CONTACT SURFACES: CLEANED & SANITIZED...   

    Latitude  Longitude                                           Location  \
0  41.945065 -87.816734  {'longitude': '41.945064857019986', 'latitude'...   
1  41.895692 -87.620143  {'longitude': '41.895692401410514', 'latitude'...   
2  41.982582 -87.708996  {'longitude': '41.98258181784537', 'latitude':...   
3  41.953378 -87.718848  {'longitude': '41.95337788158545', 'latitude':...   
4  41.793235 -87.777776  {'longitude': '41.7932347787373', 'latitude': ...   

       Facility Group  
0  childrens_services  
1          restaurant  
2          restaurant  
3          restaurant  
4             daycare

### C) Float to Integer

In [22]:
def clean_float(floatnumb):
    try :
        return int(float(floatnumb))
    except :
        return 0

In [23]:
eat_seat.Risk = eat_seat.Risk.apply(clean_float)    ##guarantees the Risk numbers to be integers

In [24]:
eat_seat.License = eat_seat.License.apply(clean_float)    ##guarantees the License numbers to be integers

In [25]:
eat_seat.Zip = eat_seat.Zip.apply(clean_float)    ##guarantees the zipcodes to be integers

In [26]:
len(eat_seat[eat_seat['Zip'] == 0])

3

The **clean_float()** function returns **0** if the zip code is not convertible into an integer. Here it means that there are 3 missing zip codes in the `Zip` column. 

In [27]:
eat_seat[eat_seat['Zip'] == 0].head()

Inspection ID       DBA Name  License  Risk             Address  Zip  \
114117        1464217  DUNKIN DONUTS  1515116     2  7545 N PAULINA ST     0   
153781        1106210  DUNKIN DONUTS  1515116     2  7545 N PAULINA ST     0   
156081         670661  DUNKIN DONUTS  1515116     2  7545 N PAULINA ST     0   

       Inspection Date        Inspection Type             Results  \
114117      2014-04-02                Canvass     Out of Business   
153781      2012-04-09  Canvass Re-Inspection                Pass   
156081      2012-02-21              Complaint  Pass w/ Conditions   

                                               Violations   Latitude  \
114117                                                NaN  42.019032   
153781  33. FOOD AND NON-FOOD CONTACT EQUIPMENT UTENSI...  42.019032   
156081  6. HANDS WASHED AND CLEANED, GOOD HYGIENIC PRA...  42.019032   

        Longitude                                           Location  \
114117 -87.673459  {'longitude': '42.01903180273219', 'latitude':...   
153781 -87.673459  {'longitude': '42.01903180273219', 'latitude':...   
156081 -87.673459  {'longitude': '42.01903180273219', 'latitude':...   

       Facility Group  
114117     restaurant  
153781     restaurant  
156081     restaurant

We can observe that the missing zip codes are all from the same esablishment. By doing a google search the zip code corresponding to the adress is easily retrieved :
*7545 N Paulina St
Chicago, IL 60626, USA*.

So that we just have to replace the missing zip codes by 60626.

In [28]:
def clean_zip(zipcode):
    if zipcode == 0 :
        return 60626
    else :
        return zipcode

In [29]:
eat_seat.Zip = eat_seat.Zip.apply(clean_zip)

### D) Community Areas

We found a file associating the chicago zipe codes and their associated community area. Using it we can create a new `Community Area` column.

In [30]:
zip_to_area = pd.read_csv('ZipCode_to_ComArea.csv',sep=',') ##creation of the dataframe
zip_to_area = zip_to_area.drop(columns = ['TOT2010'])

zip_to_area.ZipCode = zip_to_area.ZipCode.apply(clean_float) ##guarantees the zipcodes to be integers

zip_to_area = zip_to_area.groupby('ComArea')['ZipCode'].apply(list)    ##groups the zipcodes by community area number
zip_to_area = zip_to_area.reset_index()

In [31]:
zip_dic = zip_to_area.set_index('ComArea')['ZipCode'].to_dict()

In [32]:
eat_seat = newcolfromdict(eat_seat, zip_dic, 'Zip', 'Community Area')

In [33]:
eat_seat.head(3)

Inspection ID                          DBA Name  License  Risk  \
0        2320971  JUMPSTART EARLY LEARNING ACADEMY  2589822     3   
1        2320918                         BEEFSTEAK  2698445     3   
2        2320986                     BABA'S COFFEE  2423353     3   

                   Address    Zip Inspection Date Inspection Type  \
0       7559 W ADDISON ST   60634      2019-11-01         Canvass   
1       303 E SUPERIOR ST   60611      2019-11-01         License   
2  5544-5546 N KEDZIE AVE   60625      2019-11-01         Canvass   

              Results                                         Violations  \
0  Pass w/ Conditions  5. PROCEDURES FOR RESPONDING TO VOMITING AND D...   
1                Pass  39. CONTAMINATION PREVENTED DURING FOOD PREPAR...   
2            No Entry                                                NaN   

    Latitude  Longitude                                           Location  \
0  41.945065 -87.816734  {'longitude': '41.945064857019986', 'latitude'...   
1  41.895692 -87.620143  {'longitude': '41.895692401410514', 'latitude'...   
2  41.982582 -87.708996  {'longitude': '41.98258181784537', 'latitude':...   

       Facility Group Community Area  
0  childrens_services             76  
1          restaurant              8  
2          restaurant             14

### 2.3 The Violations Column

The Food Code Rules has changed since the 1st July 2018. After investigating those changes, it seems that only the denomination of the violations but not the violation itself has changed, and a few additionnal violations has been added in the possible violations. It means that those changes does not need more processing and can just be considered together as a common list of violations.

In [34]:
len(eat_seat.Violations.unique())

142356

In [35]:
eat_seat.Violations[0]

'5. PROCEDURES FOR RESPONDING TO VOMITING AND DIARRHEAL EVENTS - Comments: OBSERVED INCOMPLETE CLEAN UP KIT ON SITE.INSTRUCTED MANAGEMENT TO HAVE A COMPLETE CLEAN UP KIT BY NEXT INSPECTION.CLEAN UP PROCEDURES  ON SITE PRIORITY FOUNDATION VIOLATION  7-38-005   NO CITATION ISSUED | 10. ADEQUATE HANDWASHING SINKS PROPERLY SUPPLIED AND ACCESSIBLE - Comments: OBSERVED EXPOSED HAND SINK NOT ACCESSIBLE FOR HANDWASHING IN THE FOOD PREP AREA.INSTRUCTED MANAGEMENT TO INSTALL EXPOSED HAND SINK AT A LOCATION THAT IS ACCESSIBLE. PRIORITY FOUNDATION VIOLATION 7-38-030(c) NO CITATION ISSUED | 56. ADEQUATE VENTILATION & LIGHTING; DESIGNATED AREAS USED - Comments: OBSERVED HOT LINE HOOD FILTERS DUSTY.INSTRUCTED MANAGEMENT TO CLEAN HOT LINE HOOD FILTERS.'

It seems that every violation is a unique entry because it contains not only the violation type but also the comments of the inspectors. We have to split the Violations column into 3 different columns :
- Violation number
- Violation type
- Violation comments

It seems that every violation cell is architectured this way :
"number of the violation". "TYPE OF THE VIOLATION" - Comments : "comments of the inspector" (this format repeated as many times as the number of violations detected the day of the inspection, separated with a vertical line)

We just want to keep the violation number (because we can check which violation it is online). We create a column NumberViolations containing the ID of the violations found during the corresponding investigation. The rest is not kept because the titles of the violations can be found online and we do not plan on using the comments of the investigators.

As we can do this cleaning only for the rows where the field Violations is not empty, we will temporarily drop all the other rows.

In [36]:
temp = eat_seat.dropna(subset=['Violations'], axis = 0, how = 'all')

In [37]:
violations = temp.apply(lambda row: re.findall('\|\s([0-9]+)[.]', str(row['Violations'])), axis = 1)

In [38]:
first_violations = temp.apply(lambda row: row['Violations'].split('.')[0], axis = 1)

In [39]:
for violation, first_violation in zip(violations, first_violations):
    violation.append(first_violation)

flat_list = [item for sublist in violations for item in sublist]
unique, counts = np.unique(flat_list, return_counts=True)

In [40]:
temp = temp.assign(NumberViolations = violations)

In [41]:
temp = temp[['Inspection ID', 'NumberViolations']]
temp.head()

Inspection ID          NumberViolations
0        2320971               [10, 56, 5]
1        2320918                  [55, 39]
3        2320910              [53, 58, 51]
4        2320904                      [16]
5        2320969  [38, 55, 55, 58, 60, 38]

Now that we have a dataframe with every inspection ID of the inspections where violations has been found and a column containing the list of those violations, we can add it to the primary dataframe.

In [42]:
eat_seat = pd.merge(eat_seat, temp, how='left', on='Inspection ID', left_index=True, right_index=False)

In [43]:
eat_seat = eat_seat.set_index(['Inspection ID']) #redifines the Index
eat_seat.head()

DBA Name  License  Risk  \
Inspection ID                                                    
2320971        JUMPSTART EARLY LEARNING ACADEMY  2589822     3   
2320918                               BEEFSTEAK  2698445     3   
2320986                           BABA'S COFFEE  2423353     3   
2320910                 J.T.'S GENUINE SANDWICH  2689893     3   
2320904              KID'Z COLONY DAYCARE, INC.  2215609     3   

                               Address    Zip Inspection Date Inspection Type  \
Inspection ID                                                                   
2320971             7559 W ADDISON ST   60634      2019-11-01         Canvass   
2320918             303 E SUPERIOR ST   60611      2019-11-01         License   
2320986        5544-5546 N KEDZIE AVE   60625      2019-11-01         Canvass   
2320910             3970 N ELSTON AVE   60618      2019-11-01         License   
2320904             6287 S ARCHER AVE   60638      2019-11-01         Canvass   

                          Results  \
Inspection ID                       
2320971        Pass w/ Conditions   
2320918                      Pass   
2320986                  No Entry   
2320910                      Pass   
2320904                      Fail   

                                                      Violations   Latitude  \
Inspection ID                                                                 
2320971        5. PROCEDURES FOR RESPONDING TO VOMITING AND D...  41.945065   
2320918        39. CONTAMINATION PREVENTED DURING FOOD PREPAR...  41.895692   
2320986                                                      NaN  41.982582   
2320910        51. PLUMBING INSTALLED; PROPER BACKFLOW DEVICE...  41.953378   
2320904        16. FOOD-CONTACT SURFACES: CLEANED & SANITIZED...  41.793235   

               Longitude                                           Location  \
Inspection ID                                                                 
2320971       -87.816734  {'longitude': '41.945064857019986', 'latitude'...   
2320918       -87.620143  {'longitude': '41.895692401410514', 'latitude'...   
2320986       -87.708996  {'longitude': '41.98258181784537', 'latitude':...   
2320910       -87.718848  {'longitude': '41.95337788158545', 'latitude':...   
2320904       -87.777776  {'longitude': '41.7932347787373', 'latitude': ...   

                   Facility Group Community Area NumberViolations  
Inspection ID                                                      
2320971        childrens_services             76      [10, 56, 5]  
2320918                restaurant              8         [55, 39]  
2320986                restaurant             14              NaN  
2320910                restaurant             22     [53, 58, 51]  
2320904                   daycare             64             [16]

### 2.4 The Inspection Results

In [44]:
eat_seat.isnull().sum()

DBA Name                0
License                 0
Risk                    0
Address                 0
Zip                     0
Inspection Date         0
Inspection Type         1
Results                 0
Violations          51550
Latitude              515
Longitude             515
Location              515
Facility Group          0
Community Area          0
NumberViolations    51550
dtype: int64

We see that there are more than 50'000 rows where the Violations column is empty. We have to study wether those cells are empty because there were no violations (meaning the estalishment is healthy) or because the inspection was not successfull (meaning we can drop the row because it can not be used in our research).

## 

## ****Our research is indeed based on the study of the correlation between the number of crimes in an area and the healthiness of the establishments there. Another question that we will investigate is wether the healthiness of establishments is correlated with the owner of those establishments (aka if every establishments of an unhealthy owner are unhealthy or not)**** ##

##

In [45]:
eat_seat.Results.unique()

array(['Pass w/ Conditions', 'Pass', 'No Entry', 'Fail',
       'Out of Business', 'Not Ready', 'Business Not Located'],
      dtype=object)

We create new databases for every entries of the column Results in order to study them.

In [46]:
noentry = eat_seat[eat_seat['Results']=='No Entry']

In [47]:
outofbusiness = eat_seat[eat_seat['Results']=='Out of Business']

In [48]:
notready = eat_seat[eat_seat['Results']=='Not Ready']

In [49]:
businessnotlocated = eat_seat[eat_seat['Results']=='Business Not Located']

In [50]:
passwithconditions = eat_seat[eat_seat['Results']=='Pass w/ Conditions']

In [51]:
passed = eat_seat[eat_seat['Results']=='Pass']

In [52]:
fail = eat_seat[eat_seat['Results']=='Fail']

We investigate how many of each results' type has the column Violations empty.

In [53]:
results_dic = {'No Entry' : noentry, 'Out of Business' : outofbusiness, 'Not Ready' : notready,\
               'Business Not Located' : businessnotlocated, 'Pass With Conditions' : passwithconditions, 'Pass' : passed, 'Fail' : fail}

In [54]:
for name, result in results_dic.items() :
    print(name, ':', len(result[result['Violations'].isnull()]), 'empty Violation columns /', len(result),\
          'columns =', (len(result[result['Violations'].isnull()])/len(result)), '\n')

No Entry : 5768 empty Violation columns / 6199 columns = 0.9304726568801419 

Out of Business : 16728 empty Violation columns / 16757 columns = 0.9982693799606135 

Not Ready : 1799 empty Violation columns / 1849 columns = 0.9729583558680368 

Business Not Located : 65 empty Violation columns / 65 columns = 1.0 

Pass With Conditions : 444 empty Violation columns / 26847 columns = 0.016538160688345068 

Pass : 23676 empty Violation columns / 105356 columns = 0.2247237936140324 

Fail : 3070 empty Violation columns / 37656 columns = 0.08152751221584874 



We see that almost every entries where the Result is either 'No entry', 'Out of Business', 'Not ready' or 'Business not located' have the Violations field empty. We can safely drop those lines because they are not pertinent for our research.

In [55]:
results = ['Pass', 'Pass w/ Conditions', 'Fail']

In [56]:
eat_seat = eat_seat.loc[eat_seat['Results'].isin(results)]
len(eat_seat)

169859

Now we have to take care of the cases where there is no Violations and the Result is either Pass, Fail or Pass with conditions.

When the result is Pass and the Violation field is empty, we can add a empty list in the column "NumberViolations".

In [57]:
temp = eat_seat[eat_seat['Results'] == 'Pass']
temp = temp[temp['Violations'].isnull()]
temp.head()

DBA Name  License  Risk  \
Inspection ID                                                
2320960        CARNICERIA Y TAQUERIA TIERRA  2428138     3   
2320932         St. Mary of the Wood School  1878517     1   
2320974                              DO EAT  2492365     3   
2320850                 THE HOXTON, CHICAGO  2627738     1   
2320811                        ANDHRADARBAR  2595253     3   

                              Address    Zip Inspection Date  \
Inspection ID                                                  
2320960        3312-3314 W NORTH AVE   60647      2019-11-01   
2320932            7033 N MOSELLE ST   60646      2019-11-01   
2320974            3141 S HALSTED ST   60608      2019-11-01   
2320850               200 N GREEN ST   60607      2019-10-31   
2320811        2240-2242 W DEVON AVE   60659      2019-10-31   

                       Inspection Type Results Violations   Latitude  \
Inspection ID                                                          
2320960          Canvass Re-Inspection    Pass        NaN  41.910185   
2320932                        Canvass    Pass        NaN  42.008345   
2320974                      Complaint    Pass        NaN  41.836872   
2320850                        License    Pass        NaN  41.885699   
2320811        Complaint Re-Inspection    Pass        NaN  41.997897   

               Longitude                                           Location  \
Inspection ID                                                                 
2320960       -87.709907  {'longitude': '41.91018535990397', 'latitude':...   
2320932       -87.777039  {'longitude': '42.00834548005445', 'latitude':...   
2320974       -87.646102  {'longitude': '41.836871979178554', 'latitude'...   
2320850       -87.648789  {'longitude': '41.885699197163355', 'latitude'...   
2320811       -87.686586  {'longitude': '41.99789701348497', 'latitude':...   

              Facility Group Community Area NumberViolations  
Inspection ID                                                 
2320960           Not listed             24              NaN  
2320932               school             13              NaN  
2320974           restaurant             60              NaN  
2320850           restaurant             32              NaN  
2320811           restaurant             13              NaN

In [58]:
emptyliste = []
for i in range (0, len(temp)) :
    emptyliste.append(0)

temp['NumberViolationEmpty'] = emptyliste

In [59]:
temp.head()

DBA Name  License  Risk  \
Inspection ID                                                
2320960        CARNICERIA Y TAQUERIA TIERRA  2428138     3   
2320932         St. Mary of the Wood School  1878517     1   
2320974                              DO EAT  2492365     3   
2320850                 THE HOXTON, CHICAGO  2627738     1   
2320811                        ANDHRADARBAR  2595253     3   

                              Address    Zip Inspection Date  \
Inspection ID                                                  
2320960        3312-3314 W NORTH AVE   60647      2019-11-01   
2320932            7033 N MOSELLE ST   60646      2019-11-01   
2320974            3141 S HALSTED ST   60608      2019-11-01   
2320850               200 N GREEN ST   60607      2019-10-31   
2320811        2240-2242 W DEVON AVE   60659      2019-10-31   

                       Inspection Type Results Violations   Latitude  \
Inspection ID                                                          
2320960          Canvass Re-Inspection    Pass        NaN  41.910185   
2320932                        Canvass    Pass        NaN  42.008345   
2320974                      Complaint    Pass        NaN  41.836872   
2320850                        License    Pass        NaN  41.885699   
2320811        Complaint Re-Inspection    Pass        NaN  41.997897   

               Longitude                                           Location  \
Inspection ID                                                                 
2320960       -87.709907  {'longitude': '41.91018535990397', 'latitude':...   
2320932       -87.777039  {'longitude': '42.00834548005445', 'latitude':...   
2320974       -87.646102  {'longitude': '41.836871979178554', 'latitude'...   
2320850       -87.648789  {'longitude': '41.885699197163355', 'latitude'...   
2320811       -87.686586  {'longitude': '41.99789701348497', 'latitude':...   

              Facility Group Community Area NumberViolations  \
Inspection ID                                                  
2320960           Not listed             24              NaN   
2320932               school             13              NaN   
2320974           restaurant             60              NaN   
2320850           restaurant             32              NaN   
2320811           restaurant             13              NaN   

               NumberViolationEmpty  
Inspection ID                        
2320960                           0  
2320932                           0  
2320974                           0  
2320850                           0  
2320811                           0

In [60]:
temp = temp[['NumberViolationEmpty']]
temp.head()

NumberViolationEmpty
Inspection ID                      
2320960                           0
2320932                           0
2320974                           0
2320850                           0
2320811                           0

In [61]:
len(eat_seat)

169859

In [62]:
len(temp)

23676

In [63]:
eat_seat = pd.merge(eat_seat, temp, right_index = True, left_index = True, how = 'outer')

In [64]:
eat_seat.head(100)

DBA Name  License  Risk                Address  \
Inspection ID                                                                 
44247                      EAT A PITA  1222441     3     3155 N HALSTED ST    
44248                      LA GONDOLA  1336561     3    2914 N ASHLAND AVE    
44249            Azha Restaurant Inc.  1334073     3     960 W BELMONT AVE    
44250           SAINT JOSEPH HOSPITAL  1144381     3  2900 N LAKE SHORE DR    
44251           SAINT JOSEPH HOSPITAL  1144380     3  2900 N LAKE SHORE DR    
...                               ...      ...   ...                    ...   
54234          CHURCH'S CHICKEN # 577  1273264     2     431 N AUSTIN BLVD    
54235                    QORNER FOODS  2013363     1    5556 W HARRISON ST    
54236          MADISON FOOD MART, INC  2014259     1     5402 W MADISON ST    
54237                   MORE FOR LESS  1948557     2    4851 W CHICAGO AVE    
54238              SWAN'S FOOD & DELI  1646850     2    1325 N CENTRAL AVE    

                 Zip Inspection Date        Inspection Type Results  \
Inspection ID                                                         
44247          60657      2010-01-05              Complaint    Pass   
44248          60657      2010-01-21                Canvass    Pass   
44249          60657      2010-01-21  Canvass Re-Inspection    Pass   
44250          60657      2010-02-09                Canvass    Pass   
44251          60657      2010-02-09                Canvass    Pass   
...              ...             ...                    ...     ...   
54234          60644      2010-01-22              Complaint    Pass   
54235          60644      2010-01-22                License    Fail   
54236          60644      2010-01-25                License    Fail   
54237          60651      2010-01-25              Complaint    Pass   
54238          60651      2010-01-26                Canvass    Pass   

                                                      Violations   Latitude  \
Inspection ID                                                                 
44247          30. FOOD IN ORIGINAL CONTAINER, PROPERLY LABEL...  41.939441   
44248          32. FOOD AND NON-FOOD CONTACT SURFACES PROPERL...  41.934679   
44249          35. WALLS, CEILINGS, ATTACHED EQUIPMENT CONSTR...  41.940027   
44250          33. FOOD AND NON-FOOD CONTACT EQUIPMENT UTENSI...  41.934403   
44251          33. FOOD AND NON-FOOD CONTACT EQUIPMENT UTENSI...  41.934403   
...                                                          ...        ...   
54234          35. WALLS, CEILINGS, ATTACHED EQUIPMENT CONSTR...  41.887993   
54235          2. FACILITIES TO MAINTAIN PROPER TEMPERATURE -...  41.872948   
54236          2. FACILITIES TO MAINTAIN PROPER TEMPERATURE -...  41.880416   
54237          38. VENTILATION: ROOMS AND EQUIPMENT VENTED AS...  41.894878   
54238          37. TOILET ROOM DOORS SELF CLOSING: DRESSING R...  41.904810   

               Longitude                                           Location  \
Inspection ID                                                                 
44247         -87.649103  {'longitude': '41.93944115701468', 'latitude':...   
44248         -87.668625  {'longitude': '41.93467889167928', 'latitude':...   
44249         -87.653811  {'longitude': '41.94002687624888', 'latitude':...   
44250         -87.636806  {'longitude': '41.93440298707838', 'latitude':...   
44251         -87.636806  {'longitude': '41.93440298707838', 'latitude':...   
...                  ...                                                ...   
54234         -87.774763  {'longitude': '41.887993114472906', 'latitude'...   
54235         -87.763772  {'longitude': '41.87294753276929', 'latitude':...   
54236         -87.760279  {'longitude': '41.88041576281324', 'latitude':...   
54237         -87.747830  {'longitude': '41.89487824738007', 'latitude':...   
54238         -87.765604  {'longitude': '41.904810358201594', 'latitude'...   

              Facilit

In [66]:
eat_seat['TotViolations'] = eat_seat['NumberViolations'] + eat_seat['NumberViolationEmpty']

In [ ]:
#eat_seat.head() CA JOUE PAS MAIS L'IDEE EST LA

In [ ]:
eat_seat[eat_seat.index.isin(temp.index)]['NumberViolations']

When the result is either Fail or Pass with conditions but the Violations field is empty, we will drop those rows because there are missing values. An establishment can indeed not fail an inspection or receive conditions when no violation is found, those entries make no sense and can not be taken into account in our research.

### 3. Computation of an healthiness score per facility

In [ ]:
#keeping just one occurence of License
df_restaurants = eat_seat.drop_duplicates(subset=['License'])
df_restaurants.head()

In [ ]:
df_restaurants = df_restaurants.set_index(['License']) #redifines the Index

In [ ]:
df_restaurants.head()

In [ ]:
#drop useless columns
df_restaurants = df_restaurants.drop(columns = ['Inspection Date', 'Inspection Type', 'Results', 'NumberViolations'])
df_restaurants.head()

Now we have a dataframe with every facility listed in our preprocessed database. We want to create two columns, one containing how many times the facility has been inspected, and one containing every violation of this facility. 

In [ ]:
#compute the number of inspections in the primary dataset per license number

dic = {}

for license in df_restaurants.index :
    score = len(eat_seat[eat_seat['License'] == license])
    dic[license] = score

In [ ]:
#add this number in the new database
df_restaurants['NumberOfInspections'] = dic.values()

In [ ]:
df_restaurants.head()

In [ ]:
#create a dic with key = licenses and values = list of every violations per license
dic = {}

for license in df_restaurants.index :
    liste = []
    inspections = eat_seat['NumberViolations'][eat_seat['License'] == license]
    for inspection in inspections :
        for violation in inspection :
            liste.append(violation)
    dic[license] = liste

In [ ]:
#add it to the dataframe
df_restaurants['TotalViolations'] = dic.values()

In [ ]:
df_restaurants.head()

In [ ]:
#only keep the two new columns
df_restaurants = df_restaurants.drop(columns = ['DBA Name', 'Risk', 'Address', 'Zip', 'Latitude', 'Longitude', 'Facility Group', 'Community Area'])
df_restaurants.head()

Another step would be to add a ponderation to the violations depending on the gravity of the violation but it will maybe be done later. For now, we compute the score of a facility by multiplicating the number of violations by the number of inspections :

In [ ]:
df_restaurants['score'] = df_restaurants.apply(lambda x : x['NumberOfInspections'] * len(x['TotalViolations']))
df_restaurants

In [ ]:
eat_seat = pd.merge(eat_seat, df_restaurants, how='left', on='License', left_index=True, right_index=False)

In [ ]:
eat_seat.head()

### 4.1 Additional Dataset - BUSINESS LICENSES/OWNERS

We found two datasets on Kaggle gathering the business licenses and the business owners of Chicago. It could be interesting to observe the results of different establishments owned by the same person.

### A) Business Licenses

The first dataset contains the details about every licensed establishments. There are a lot of columns but the only ones interesting us are :
- the `license id` column to have a link with the *chicago food inspections* dataset
- the `account number` column to have a link with the *business owners* dataset
- the `police district` column in case we want to have a link with the *crime* dataset

In [ ]:
licenses = pd.read_csv('business-licenses.csv', sep=',') #creation of the dataframe
licenses = licenses.rename(str.lower, axis='columns')
licenses = licenses.drop(columns = ['city', 'state', 'id', 'precinct', 'ward precinct', 'business activity id',\
                                'license number', 'application type', 'application created date',\
                                'application requirements complete', 'payment date', 'conditional approval',\
                                'license term start date', 'license term expiration date', 'license approved for issuance',\
                                'date issued', 'license status', 'license status change date', 'ssa',\
                                'historical wards 2003-2015', 'zip codes', 'wards', 'census tracts', 'location',\
                                'license code', 'license description', 'business activity', 'site number',\
                               'zip code', 'latitude', 'longitude', 'address', 'legal name', 'doing business as name',\
                                'community areas', 'ward'])
licenses['police district'] = licenses['police district'].apply(clean_float) 

licenses.head(3)

In [ ]:
licenses = licenses.set_index('account number')
licenses.head(3)

In [ ]:
print(len(licenses.index))

### B) Business Owners

The second dataset contains the details about every license owners. We have decided to keep the following columns :
- the `account number` column to have a link with the *business licenses* dataset
- the `owner first name` and the `owner last name` columns in order to create a `full name` column

In [ ]:
owners = pd.read_csv('business-owners.csv',sep=',') #creation of the dataframe
owners = owners.rename(str.lower, axis='columns')
owners = owners.drop(columns = ['suffix', 'legal entity owner', 'owner middle initial', 'legal name', 'title'])
owners.head(3)

In [ ]:
owners['full name'] = owners['owner first name'] + ' ' + owners['owner last name']
owners = owners.drop(columns = ['owner first name', 'owner last name'])

A `full name` column is enough for the needs that we have to link the licenses to the owners.

In [ ]:
owners.head(3)

In [ ]:
len(owners['account number'].unique())

In [ ]:
len(owners['full name'].unique())

Here we can see that the number of accounts is not the same that the number of full names. For now, we will consider that a same account can be shared by several people (for example, it could be the case for partners owning a business together).

In [ ]:
owners = pd.DataFrame(owners.groupby('account number')['full name'].apply(list))

In [ ]:
owners.head()

We can observe that the lists of the `full name` column contain duplicates and 'nan' values. A function *clean_list* can be defined to clean them.

In [ ]:
def clean_list(liste) :
    cleaned = []
    for element in liste :
        if type(element) == str and element not in cleaned :
            cleaned.append(element)
    return cleaned

In [ ]:
owners['full name'] = owners['full name'].apply(clean_list)

In [ ]:
owners.head()

### C) Business Licenses-Owners

Setting both the indexes of the *licenses* and the *owners* dataframes we can now merge them together.

In [ ]:
business = pd.merge(licenses, owners, right_index = True, left_index = True)

In [ ]:
business.head()

In [ ]:
business.reset_index(level=0, inplace=True)

In [ ]:
business.head()

### D) Second Main Dataframe

Setting both the indexes of the *business* and the *eat_seat* dataframes we can now merge them together.

In [ ]:
business = business.rename(columns= {'license id' : 'License'})
business = business.set_index('License')

eat_seat = eat_seat.set_index('License')

In [ ]:
eat_seat_2 = pd.merge(eat_seat, business, right_index = True, left_index = True)

In [ ]:
eat_seat_2.head()

In [ ]:
eat_seat_2.to_csv('newfood.csv')

### 4.2 Additional Dataset - CRIME 

## looking at the dataset and modify it

In [ ]:
crime_2001_2004 = pd.read_csv('Chicago_Crimes_2001_to_2004.csv', error_bad_lines=False)

In [ ]:
crime_2005_2007 = pd.read_csv('Chicago_Crimes_2005_to_2007.csv', error_bad_lines=False)

In [ ]:
crime_2008_2011 = pd.read_csv('Chicago_Crimes_2008_to_2011.csv', error_bad_lines=False)

In [ ]:
crime_2012_2017 = pd.read_csv('Chicago_Crimes_2012_to_2017.csv', error_bad_lines=False)

These 4 cells load the data taken from kaggle in a panda DataFrame

In [ ]:
crime = crime_2001_2004.append(crime_2005_2007).append(crime_2008_2011).append(crime_2012_2017)

crime.head()

This cell is uniting all the data in a single DataFrame to make the data cleaning easier

In [ ]:
crime = crime.drop(['Unnamed: 0','Case Number','IUCR','Arrest', 'Domestic','FBI Code','X Coordinate','Y Coordinate','Updated On', 'Location Description'], axis=1)

crime.head()

Here the columns that are not useful for our analysis are dropped

In [ ]:
Block_map = []
for x in crime['Block'] :
    y =x.replace('XX','00')
    Block_map.append(y)

The Adress under the `Block` column are anonimized with the XX at the end of the adress number. We'll use the adress for our map so this is a problem. We can't have the exact adress, so we make the aproximation of using the 00 adress instead of the XX. This is not exact but will give an approximative location of where the crimes where comitted.

In [ ]:
crime['Block_mapable'] = Block_map

crime.head()

The new Serie created with the 00 instead of the XX is load in the DataFrame

In [ ]:
for x in crime.columns :
    print(x + ' : ' + str(crime[x].isnull().values.any()) + ' --> ' + str(crime[x].isnull().sum()))

Here we are looking at missing data, to see there number and position, and hen be able to find a way to deal with them.

We can see that most of the missing data are set in the `Ward`and `Community Area` columns.

Most of the missing data are set in the location part of the data

By now, we keep all the data and we will see later if we have to deal with missing values (maybe for the map).

In [ ]:
print(set(crime['Primary Type']))

We print all the Primary Types of crimes to create a dictionary with every crime type keyed with the minimal number of year of prison that someone risk when commiting this crime. More details below

In [ ]:
crime_penalty = {'PROSTITUTION' : 1, 'DOMESTIC VIOLENCE' : 1, 'MOTOR VEHICLE THEFT' : 3, 'ASSAULT' : 1, 'OFFENSE INVOLVING CHILDREN' : 1,\
                 'RITUALISM' : 1, 'BATTERY' : 1, 'CRIM SEXUAL ASSAULT' : 4, 'GAMBLING' : 1,\
                 'PUBLIC INDECENCY' : 1, 'OTHER OFFENSE' : 1, 'LIQUOR LAW VIOLATION' : 1, 'OTHER NARCOTIC VIOLATION' : 1, 'OBSCENITY' : 1,\
                 'NON-CRIMINAL' : 1, 'KIDNAPPING' : 3, 'HOMICIDE' : 20, 'NARCOTICS' : 1, 'ARSON' : 6, 'DECEPTIVE PRACTICE' : 1, 'ROBBERY' : 3,\
                 'BURGLARY' : 3, 'INTIMIDATION' : 2, 'HUMAN TRAFFICKING' : 4, 'SEX OFFENSE' : 4, 'CRIMINAL TRESPASS' : 1,\
                 'CONCEALED CARRY LICENSE VIOLATION' : 2, 'CRIMINAL DAMAGE' : 1, 'INTERFERENCE WITH PUBLIC OFFICER' : 1, 'PUBLIC PEACE VIOLATION' : 1,\
                 'WEAPONS VIOLATION' : 1, 'THEFT' : 1, 'STALKING' : 1}

To make the crime_penalty dict, we look for each crime type what is the minimum prison penalty (in year). For the crime where the penalty is below 1 year of prison, we took the 1 value.

These values are taken from the Illinois penalty code (found on internet).

In [ ]:
Primary_Type = []
for x in crime['Primary Type'] :
    if x == 'NON-CRIMINAL (SUBJECT SPECIFIED)' or x == 'NON - CRIMINAL' :
        Primary_Type.append('NON-CRIMINAL')
    else :
        Primary_Type.append(x)
        
crime['Primary_Type'] = Primary_Type

Here we group all the NON-CRIMINAL values under one label (3 label before)

In [ ]:
crimescore = []
for x in crime['Primary_Type'] :
    crimescore.append(crime_penalty[x])

Here we create a new column with the crimescore for each row 8associated with the Primary type crime

In [ ]:
crime['crimescore'] = crimescore

In [ ]:
crime.head()

In [ ]:
crimetest = pd.DataFrame()
crimetest['Community Area'] = crime['Community Area']
crimetest['crimescore'] = crime['crimescore']

crimetest.head()

Here we select the community area and the crimescore area to calculate the crimescore for each community area

In [ ]:
crime_CA = crimetest.groupby('Community Area').sum()

crime_CA

Here we have the DataFrame interesting us : 

- Community area | Crimescore

With this df, we'll be able to see if there is a correlation between restaurant quality and crime in Chicago

# About this notebook

We saw that the data are not that hard to handle : The data is not that big, and the problems on the format of the different columns is not a problem anymore. However, the missing values on the Wards, the Community Area and the location (long, lat) is still a problem.

This can become a problem, these data without any location information are difficult to place on a map, we will certainly use them during the analysis and plotting phase, but we'll certainly delete them when we'll pin the data up on the map.

With these 3 datasets, we want to compare the global healthiness of public (hospital, ...) facilities vs private facilities (restaurant, ...).

We also want to see if there is differences of the global healthiness of facilites between all the community area. We'll then pin the facilities on a map and delimit healthy and unhealthy community area.

We'll next watch if there are similarities in the healthiness of several restaurants directed by one person (food chain), and if there is a difference between the restaurant of people that handle only one restaurant vs people that handle several restaurants.

We also want to see if there is a correlation between the healthiness of the restaurant and the crimes that take place in the same community area.

All these correlations, differences, similarities will be visualized and mapped.

To compare the healthiness of restaurants (and community area) and the criminality in different community area, we designed a way to get the healthiness score and the crimescore of these different things.

We calculated the healthiness score by taking the several infractions, and the score given in the dataset (to be more precise).

We calculated the crime score based on the minimal prison penalty of each infraction. We took the primary infractions only, this was easier to set up, but can bring some little imprecisions.

# Can the data bring us answer to our questions ?

We think so, All the informations we need to make the different analysis are in the 3 datasets. For the healthiness of the restaurants and in the different community area, all the data are in the food inspections dataframe.

For the analysis about the different owners of different facilities, all the informations needed were processed and are in the license dataset.

And for the crime related analysis, the informations were processed and we have all we need in the crime dataset.

# What comes next

We'll first start to get deeper into the analysis, and maybe come back to modify or arrange the data one last time according to our needs.

When the analysis will be completed, we'll start to get to the visualization : What is the best way to present our results ? Is the map still a good visualization ? How to handle the map ?

Then we'll finally report our results, and start preparing the presentation.